gen/qr.hoon

```diff
@@ -0,0 +1,359 @@
+  ^-  grid
+  =/  n=@u  (lent qr)
+  =/  i=@u  0
+  :-  (reap (mul 2 (add 2 n)) %.n)
+  |-
+  ?:  =(i n)
+    :-  (reap (mul 2 (add 2 n)) %.n)  ~
+  :-  (zing [[%.n %.n ~] (snag i qr) [%.n %.n ~] ~])
+  $(i (add 1 i))
+::
+++  add-mask-information
+  |=  [mask=@u qr=grid]
+  ^-  grid
+  =/  info  (format-strings mask)
+  =/  i=@u  0  :: TODO dont hardcode mask 000
+  |-
+  ?:  =(i (lent info))
+    qr
+  ?:  =((snag i info) 1)
+    =/  coords-one  (snag i info-route-one)
+    =/  coords-two  (snag i info-route-two)
+    %-  tm  :-  [-.coords-one +.coords-one]
+    %-  tm  :-  [-.coords-two +.coords-two]
+    $(i (add 1 i))
+  $(i (add 1 i))
+::
+++  pick-mask
+  |=  qr=grid
+  ^-  [@u grid]
+  =/  i=@u  0
+  :-  0  :: TODO hardcoding mask 000 now. implement other masks + penalty scoring
+  |-
+  ?:  =(i (lent data-route))
+    qr
+  =/  coords  (snag i data-route)
+  =/  x  -.coords
+  =/  y  +.coords
+  ?:  =(0 (mod (add x y) 2))  :: mask 000
+    %-  tm  :-  [x y]
+    $(i (add 1 i))
+  $(i (add 1 i))
+::
+
+++  atm  :: TODO surely there's a cleaner way to do this too? (+turn...)
+  |=  [coords=(list [@u @u]) b=grid]
+  ^-  grid
+  |-
+  ?~  coords  b
+  %+  tm  i.coords
+  $(coords t.coords)
+::
+++  tm  :: TODO surely there's a cleaner way to do this?
+  |=  [coord=[@u @u] b=grid]
+  ^-  grid
+  =/  c=@u  -.coord
+  =/  r=@u  +.coord
+  =/  count=@u  0
+  =/  s=@u  (mul 2 c)  :: TODO yikes module abstraction needed
+  |-
+  ?:  =(count (lent b))
+    ~
+  =/  row=(list ?)  (snag count b)
+  :-  ?:  =(count r)
+         (zing ~[(scag s row) ~[!(snag s row) !(snag (add 1 s) row)] (slag (add 2 s) row)])
+    row
+  $(count (add 1 count))
+::
+::::::::::::::::::::::::::::::::::::::::::::::::5: message encoding helpers
+::
+++  add-error-correction
+  |=  msg=tape
+  ^-  (list @)
+  =/  cod  (msg-to-codewords msg)
+  =/  pol  (codewords-to-polynomial cod)
+  (byte-encode (synthetic-div [pol generator-polynomial]))
+::
+++  add-data
+  |=  [data=(list @) qr=grid]
+  ^-  grid
+  =/  i=@u  0
+  |-
+  ?:  =(i (lent data))
+    qr
+  ?:  =((snag i data) 1)
+    =/  coords  (snag i data-route)
+    %-  tm  :-  [-.coords +.coords]
+    $(i (add 1 i))
+  $(i (add 1 i))
+::
+++  padding
+  |=  char=@tD
+  ^-  (list @)
+  =/  binary-list-flipped  (rip 0 char)
+  =/  binary-list  (flop binary-list-flipped)
+  =/  padded  (zing [(reap (sub 8 (lent binary-list)) 0) binary-list ~])
+  padded
+::
+++  mode-indicator
+  ^-  (list @)
+  [0 1 0 0 ~] :: alphanumeric mode
+::
+++  char-count-indicator
+  |=  l=tape
+  ^-  (list @)
+  %-  padding  (lent l)
+::
+++  byte-encode
+  |=  m=(list @)
+  ^-  (list @)
+  =/  i=@u  0
+  %-  zing
+  |-
+  ?:  =(i (lent m))
+    ~
+  :-  (padding (snag i m))
+  $(i (add 1 i))
+::
+++  add-padding-zeros
+  :: pad a string of bits to ensure it is divisble by 8
+  |=  wip=(list @)
+  ^-  (list @)
+  %-  zing
+  :-  wip
+  :-  (reap (sub (lent wip) (mul 8 (div (lent wip) 8))) 0)   :: make sure this is an evenly divisible by 8
+  ~
```

Comment:
In cases like this, where you have a list or similarly simple structure, with a complicated value in the middle, I personally like using `=-`. Here, that would look something like:

```hoon
%-  zing
=-  ~[wip (reap - 0)]  ::NOTE  equivalent to [wip (reap - 0) ~]
%+  sub  (lent wip)
(mul 8 (div (lent wip) 8))
```

Reply:
Isn't this just `(sub 8 (mod (lent wip) 8))`? Also `(zing wip foo ~)` is just `(weld wip foo)` :P

Reply:
Hah, that's a good catch. (: